sql/core/src/test/scala/org/apache/spark/sql/ProductAggSuite.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.functions.{ col, lit, product }
+import org.apache.spark.sql.test.SharedSparkSession
+
+
+class ProductAggSuite extends QueryTest with SharedSparkSession {
+  // Sequence of integers small enough that factorial is representable exactly as DoubleType:
+  private lazy val data16 = spark.range(1, 17).toDF("x")
+
+  private lazy val factorials = (1 to 16).scanLeft(1L) { case (f, x) => f * x }
+
+  test("bare factorial") {
+    implicit val enc = Encoders.scalaDouble
+
+    val prod = data16.agg(product(col("x"))).as[Double].head
+    val expected = (1L to 16L).reduce { _ * _ }.toDouble
+
+    assert(prod === expected)
+    assert(prod === factorials(16))
+  }
+
+  test("type flexibility") {
+    import org.apache.spark.sql.types.{ ByteType, DoubleType, FloatType, IntegerType, ShortType }
```

Comment:
Can we import on the top?  also -> `{ByteType, DoubleType, FloatType, IntegerType, ShortType}`

Reply:
Now moved to top of file.